### DISCUSSION TITLE
Output out of order when yielding from a partial rendered by Phlex

### DISCUSSION THREAD

**Discussion by adamlogic** [yippee-fun/phlex#592] [Help] [answered]
That title is a bit of a mouthful, but here's a minimal example.

I have a Phlex component that's only job is to render a partial with the provided block. The partial renders some debugging output and yields to the block.

```ruby
# home_component.rb

class HomeComponent < ApplicationComponent
  def template(&)
    render "home_partial", &
  end
end
```

```erb
<%# _home_partial.html.erb %>

<p><code>home_partial</code> before yield</p>
<%= yield %>
<p><code>home_partial</code> after yield</p>
```

Then I have an ERB view that renders the partial directly and also renders the component:

```erb
<div class="prose">
  <h2><code>home_partial</code> rendered in ERB view:</h2>

  <%= render "home_partial" do %>
    <p>content passed to home partial</p>
  <% end %>

  <hr>

  <h2><code>home_partial</code> rendered in Phlex component:</h2>

  <%= render HomeComponent.new do %>
    <p>content passed to home component</p>
  <% end %>
</div>
```

When I view this in a browser, rendering the partial directly works as I'd expect, but rendering the partial within the Phlex component has the output out of order:

![CleanShot 2023-10-17 at 15 04 21@2x](https://github.com/phlex-ruby/phlex/assets/372/ca1ef903-e8e3-40db-933b-ca8ba4c37dc3)

I understand that these are two different `render` methods. Calling `render` from ERB will return a string, and ERB is responsible for appending that string into the output buffer. Calling `render` from a Phlex component immediately appends the result into the output buffer. That doesn't explain why the `yield` behaves differently, though. Or at least I'm not seeing it.

My goal is to gradually integrate Phlex into my app, and this means I'll have lots of Phlex components rendering existing ERB partials and vice versa. My understanding from the docs was that this is supported. Am I doing something wrong?

**Comment by joeldrapper** [comment]
Hey @adamlogic, thanks for the detailed bug report and reproduction steps. You’re not doing anything wrong as far as I can tell, looking at this. It does seem to be a bug.

**Reply by joeldrapper** [reply]
@adamlogic I’ve opened an issue for this here. https://github.com/phlex-ruby/phlex-rails/issues/112

So far, I’ve been able to reproduce the issue and find exactly where it’s coming from.

When we call the `template` method on the component, we don’t actually pass it the raw block that the component was originally rendered with. Instead, [we pass it a new block](https://github.com/phlex-ruby/phlex/blob/5a4002f906dbf20e09fd087f51a5841e85187064/lib/phlex/sgml.rb#L113-L119) that calls `yield_content` with the original block. This `yield_content` method is how Phlex can treat `{ "Hello" }` like `{ plain "Hello" }`.

Here's how it works:

1. first, it gets the current size of the buffer;
2. then, it yields the original block, saving the return value for later;
3. finally, it checks if the size of the output buffer changed when yielding the block;
4. if yielding the block changed the size of the output buffer, the block must have contained an output method such as a tag, e.g. `{ input(type: "text") }`; otherwise, if the output buffer’s size hasn’t been changed by yielding the block, there’s a good chance we’re actually interested in the return value of that block, so we pass the return value to `__text__` (which is a private method similar to `plain` but it won’t raise if it’s passed something that doesn’t make sense as plain text).

Here’s the full `yield_content` method

```ruby
def yield_content
  return unless block_given?

  target = @_context.target

  original_length = target.length
  content = yield(self)
  __text__(content) if original_length == target.length

  nil
end
```

there's also a `yield_content_with_args` method, which is almost exactly the same, but it yields params, so it's ever so slightly slower.

Anyway, back to the issue: the problem is when the `template` method picks up the block in the component, it’s got this `yield_content` behaviour built in. So when the partial eventually yields it, that `yield_content` method immediately pushes the content to the buffer, which is not what you expect to happen in a partial.

Hope that makes sense.

I’m not 100% sure how to fix it yet, but I’m sure I’ll find a way. There are many other places where Phlex wanting to push to the buffer comes into conflict with ERB wanting to work with returned string values.

**Reply by joeldrapper** [reply, marked as answer]
Okay, I’ve got a fix here using `capture` to render the block into a string while in Phlex context before passing it down to another Rails renderable. https://github.com/phlex-ruby/phlex-rails/pull/113

You should be able test it by setting this in your `Gemfile`:

```ruby
gem "phlex-rails", git: "https://github.com/phlex-ruby/phlex-rails.git"
```

Do you mind if I use your example code as a test case?

**Reply by joeldrapper** [reply]
> That title is a bit of a mouthful

It was even more specific than that. I titled the PR: **“Fix out of order rendering when yielding from a partial rendered with a block delegated from a component’s content block”** 😅

**Reply by adamlogic** [reply]
What a response!! Thanks so much, not just for the fix, but for taking the time to explain what was happening. Truly above and beyond my expectations.

I've confirmed that the fix works for me, and of course you can use my example code as a test case. :-)

**Comment by mhenrixon** [comment]
I have the same issue in a different scenario: 

```html
<!-- Response -->
<!-- Before Messages::ShowView -->
<!-- Before ModalComponent -->
<dialog id="modal_message_987" class="modal" data-controller="dialog" data-action="click-&gt;dialog#clickOutside">
    <div class="modal-box w-full max-w-3x">
        <form method="dialog">
            <button class="btn btn-sm btn-circle absolute right-4 top-4" data-action="dialog#close">✕</button>
        </form>
        <h3 class="font-bold text-lg">Angustus compono amaritudo.</h3>
        <p class="py-4">Vesco tactus suasoria. Cohaero atqui antepono. Eius praesentium sto.

        Architecto comes vereor. Viduata artificiose depono. Sint tenus summisse.

        Tubineus armo vesica. Ancilla ab aut. Vorago nobis uredo.</p>
        <div class="modal-action">
            <form method="dialog">
                <!-- Before Modal::ActionComponent -->
                <a href="#" class="btn btn-primary mr-2" message-id="987" data-turbo-command="MessageCommand#deliver">Deliver</a>
                <!-- Before Modal::ActionComponent -->
                <a href="#" class="btn btn-error" message-id="987" data-turbo-command="MessageCommand#destroy">Destroy</a>
            </form>
        </div>
    </div>
</dialog>
<turbo-stream action="invoke" target="DOM">
    <template>{"id":"122a5b2b-c0ea-43ec-8c39-84fe690841be","selector":"#turbo-boost","method":"morph","args":["\u003cmeta id=\"turbo-boost\" name=\"turbo-boost\" content=\"IllDQklUaGYwSWlrcnVMQlEi--6a413c759d3ea9609f6d5514ca086b1be99c32abd611a3531a83110ba87a7975\" data-busy=\"false\" data-state=\"e30\" /\u003e"],"delay":0}</template>
</turbo-stream>
```

For the following code:

```ruby
# frozen_string_literal: true

class Messages::ShowView < ApplicationView
  include Turbo::FramesHelper

  def initialize(message:)
    @message = message

    super()
  end

  def template
    turbo_frame_tag(:modal) do
      _(
        :modal,
        id: dom_id(@message, :modal),
        title: @message.subject,
        body: @message.body,
        box_class: "w-full max-w-3x"
      ) do |c|
        c.with_action(
          title: "Deliver",
          href: "#",
          class: "btn btn-primary mr-2",
          message_id: @message.id,
          data: {
            turbo_command: "MessageCommand#deliver"
          }
        )

        c.with_action(
          title: "Destroy",
          href: "#",
          class: "btn btn-error",
          message_id: @message.id,
          data: {
            turbo_command: "MessageCommand#destroy"
          }
        )
      end
    end
  end
end
```


The `_` method is just a shorter way for me to render a component: 

```ruby
module ComponentHelper
  def component(component_name, ...)
    component_class = "#{component_name.to_s.camelize}Component".constantize
    render component_class.new(...)
  end
  alias _ component
end
```

The modal component looks like: 

```ruby
# frozen_string_literal: true

class ModalComponent < ApplicationComponent
  include Phlex::DeferredRender

  attr_accessor :id, :title, :body, :options

  def initialize(id:, title:, body:, **options)
    @id      = id
    @title   = title
    @body    = body
    @options = options
    @actions = []

    super()
  end

  def template
    dialog(id: id, class: "modal", data: { controller: "dialog", action: "click->dialog#clickOutside" }) do
      div(class: class_names("modal-box", options[:box_class])) do
        form(method: "dialog") do
          button(class: "btn btn-sm btn-circle absolute right-4 top-4", data: { action: "dialog#close" }) do
            "✕"
          end
        end

        h3(class: "font-bold text-lg") { @title }
        p(class: "py-4") { @body }

        div(class: "modal-action") do
          form(method: "dialog") do
            @actions.each do |action|
              render(action)
            end
          end
        end
      end
    end
  end

  def with_action(...)
    @actions << Modal::ActionComponent.new(...)
  end
end
```

Not entirely sure what is going wrong but this is the only thing blocking me from going hundred percent phlex.

**Reply by joeldrapper** [reply]
@mhenrixon The issue is you're using a helper from `Turbo::FramesHelper` which isn't adapted to work with Phlex. Instead, you should include `Phlex::Rails::Helpers::TurboFrameTag`.

Just replace this line like this
```diff
-  include Turbo::FramesHelper
+  include Phlex::Rails::Helpers::TurboFrameTag
```

**Reply by mhenrixon** [reply]
Didn't work: 

```ruby
def template
  turbo_frame_tag(:modal) do
    render ModalComponent.new(
      id: dom_id(@message, :modal),
      title: @message.subject,
      body: @message.body,
      box_class: "w-full max-w-3x"
    )
  end
end
```

Gives me: 

```html
<!-- Before Messages::ShowView -->
<!-- Before ModalComponent -->
<dialog id="modal_message_981" class="modal" data-controller="dialog" data-action="click-&gt;dialog#clickOutside">
    <div class="modal-box w-full max-w-3x">
        <form method="dialog">
            <button class="btn btn-sm btn-circle absolute right-4 top-4" data-action="dialog#close">✕</button>
        </form>
        <h3 class="font-bold text-lg">Amaritudo vir tendo.</h3>
        <p class="py-4">Comitatus soleo minus. Apud et toties. Perferendis demens crapula.

        Curiositas consuasor desidero. Odit sint cubitum. Arx ustulo dicta.

        Compello vaco astrum. Cupiditas defleo custodia. Degusto vociferor utrum.</p>
        <div class="modal-action">
            <form method="dialog">
                <!-- Before Modal::ActionComponent -->
                <a href="#" class="btn btn-primary float-right" message-id="981" data-turbo-command="MessageCommand#deliver">Deliver</a>
                <!-- Before Modal::ActionComponent -->
                <a href="#" class="btn btn-error float-left" message-id="981" data-turbo-command="MessageCommand#destroy">Destroy</a>
            </form>
        </div>
    </div>
</dialog>
<turbo-frame id="modal">
    <turbo-stream action="invoke" target="DOM">
        <template>{"id":"04439ecc-a094-404c-bb8e-a31952b94cd5","selector":"#turbo-boost","method":"morph","args":["\u003cmeta id=\"turbo-boost\" name=\"turbo-boost\" content=\"IkZ0S0NKenRQT2ZzRVBQRFki--9977f525fb68bbb2c1d63193049d85be9fda05845d6593ab065a987e5f2ddada\" data-busy=\"false\" data-state=\"e30\" /\u003e"],"delay":0}</template>
    </turbo-stream>
</turbo-frame>
```

While:

```ruby
def template
  turbo_frame_tag(:modal, lambda do
    render ModalComponent.new(
      id: dom_id(@message, :modal),
      title: @message.subject,
      body: @message.body,
      box_class: "w-full max-w-3x"
    )
  end)
end
```

Gives me:

```html
<!-- Before Messages::ShowView -->
<turbo-frame id="modal_#&lt;Proc:0x00000002b1e50048 /Users/mhenrixon/Code/mhenrixon/cosmos/app/views/messages/show_view.rb:14 (lambda)&gt;">
    <turbo-stream action="invoke" target="DOM">
        <template>{"id":"2a111895-970d-4800-80d3-70cf7b09d87e","selector":"#turbo-boost","method":"morph","args":["\u003cmeta id=\"turbo-boost\" name=\"turbo-boost\" content=\"Ik1sOUNQUGtPOXgwRzRTeloi--1ac26055f2a6fd8e215f8c52fb0be8e8cb600dcefcd3410755b234df9e68ba4f\" data-busy=\"false\" data-state=\"e30\" /\u003e"],"delay":0}</template>
    </turbo-stream>
</turbo-frame>
```

Something is still amiss and nested blocks were never my strong suite!

**Reply by joeldrapper** [reply]
Oh that’s very strange. It looks like it's calling `to_s` on an object. Do you think you could set up a minimal reproduction in a new rails app?

**Reply by mhenrixon** [reply]
> Do you think you could set up a minimal reproduction in a new rails app?

That's a fair ask, but I am not sure I can. First, I'll remove everything that isn't necessary. 

Could it be that I am using an erb layout?

**Reply by joeldrapper** [reply]
Wait, I think you're using the helper wrong. You should pass it a block not a lambda.

**Reply by joeldrapper** [reply]
Try this:

```ruby
def template
  turbo_frame_tag(:modal) do
    render ModalComponent.new(
      id: dom_id(@message, :modal),
      title: @message.subject,
      body: @message.body,
      box_class: "w-full max-w-3x"
    )
  end
end
```

**Reply by mhenrixon** [reply]
I have tried reproducing it in a separate repo, but the internet won't install the dependencies! I can't even try. 🤣

I'll have to come back to it later. If you have any bright ideas let me know.

**Reply by joeldrapper** [reply]
No problem, I think I found the issue above. I’m surprised passing a lambda in even worked. It's not a documented part of the `turbo_frame_tag` API. The adapter in `Phlex::Rails::Helpers::TurboFrameTag` only adapts the block. It doesn't check if the second argument is a Proc and try to adapt that.

Alternatively, placing an `&` behind your lambda when you pass it in would coerce it into a block argument, which should also do the trick.

**Reply by mhenrixon** [reply]
@joeldrapper I try like you said and get the following response: 

```html
<!-- Before Messages::ShowView -->
<!-- Before ModalComponent -->
<dialog id="modal_message_975" class="modal" data-controller="dialog" data-action="click-&gt;dialog#clickOutside">
    <div class="modal-box w-full max-w-3x">
        <form method="dialog">
            <button class="btn btn-sm btn-circle absolute right-4 top-4" data-action="dialog#close">✕</button>
        </form>
        <h3 class="font-bold text-lg">Cupiditate asporto socius.</h3>
        <p class="py-4">Theologus speciosus cogito. Sum carus brevis. Statim acsi cito.

        Aqua defero spoliatio. Amplitudo vobis rerum. Angelus crinis cursus.

        Confero quia tenax. Vobis arbitro clamo. Ullam vociferor arguo.</p>
        <div class="modal-action">
            <form method="dialog">
                <!-- Before Modal::ActionComponent -->
                <a href="#" class="btn btn-primary float-right" message-id="975" data-turbo-command="MessageCommand#deliver">Deliver</a>
                <!-- Before Modal::ActionComponent -->
                <a href="#" class="btn btn-error float-left" message-id="975" data-turbo-command="MessageCommand#destroy">Destroy</a>
            </form>
        </div>
    </div>
</dialog>
<turbo-frame id="modal">
    <turbo-stream action="invoke" target="DOM">
        <template>{"id":"a7659deb-f133-4035-b416-db3f89db9c33","selector":"#turbo-boost","method":"morph","args":["\u003cmeta id=\"turbo-boost\" name=\"turbo-boost\" content=\"IjdEUGQ4WFVtejBMcW1jSngi--7a12b1d52f03257ec1f1f6059df483872db81761af442c9f3e77c8bcdeeb4e2d\" data-busy=\"false\" data-state=\"e30\" /\u003e"],"delay":0}</template>
    </turbo-stream>
</turbo-frame>
```

For the following code: 

```ruby
def template
  turbo_frame_tag(:modal) do
    render ModalComponent.new(
      id: dom_id(@message, :modal),
      title: @message.subject,
      body: @message.body,
      box_class: "w-full max-w-3x",
    ) do |c|
      c.with_action(
        title: "Deliver",
        href: "#",
        class: "btn btn-primary float-right",
        message_id: @message.id,
        data: {
          turbo_command: "MessageCommand#deliver",
        },
      )

      c.with_action(
        title: "Destroy",
        href: "#",
        class: "btn btn-error float-left",
        message_id: @message.id,
        data: {
          turbo_command: "MessageCommand#destroy",
        },
      )
    end
  end
end
```

**Reply by mhenrixon** [reply]
There is also nothing I can do for the modal component. I tried to render it to string and and then use the string in the block with a plain but the response still comes out with the dialog first and outside of the turbo frame.


It is like it appends each thing separately to the response.

**Reply by mhenrixon** [reply]
I also verified that turbo boost isn't the culprit by removing every trace of turbo boost and then I get an empty turbo frame with id="modal".

**Reply by joeldrapper** [reply]
Hmm, this shouldn't be possible, but let's break it down and see if we can find where the issue is.

Let's skip the adapter for now and use the rails helper directly.

```ruby
def template
  helpers.turbo_frame_tag(:modal) do
    capture do
      render ModalComponent.new(
        id: dom_id(@message, :modal),
        title: @message.subject,
        body: @message.body,
        box_class: "w-full max-w-3x",
      ) do |c|
        c.with_action(
          title: "Deliver",
          href: "#",
          class: "btn btn-primary float-right",
          message_id: @message.id,
          data: {
            turbo_command: "MessageCommand#deliver",
          },
        )
  
        c.with_action(
          title: "Destroy",
          href: "#",
          class: "btn btn-error float-left",
          message_id: @message.id,
          data: {
            turbo_command: "MessageCommand#destroy",
          },
        )
      end
    end
  end
end
```

What does this output?

Make sure you're on the latest version of `phlex-rails` — run `bundle update phlex-rails`. Also, check you haven’t included something that’s overridden `capture`. I’ve [seen](https://github.com/phlex-ruby/phlex-rails/issues/141#issuecomment-1911336280) that break things before.

**Reply by mhenrixon** [reply]
I removed the DeferredRenderer for now and simplified to the following: 

```ruby
def template
  helpers.turbo_frame_tag(:modal) do
    capture do
      render ModalComponent.new(
        id: dom_id(@message, :modal),
        title: @message.subject,
        body: @message.body,
        box_class: "w-full max-w-3x",
      ) do |c|
        a(href: "#",
          class: "btn btn-primary float-right",
          message_id: @message.id,
          data: {
            turbo_command: "MessageCommand#deliver",
          }) { "Deliver" }

        a(
          href: "#",
          class: "btn btn-error float-left",
          message_id: @message.id,
          data: {
            turbo_command: "MessageCommand#destroy",
          },
        ) { "Destroy" }
      end
    end
  end
end
```

I also checked that I am on the latest version of phlex-rails and I still get the same result: 

![Uploading CleanShot 2024-01-31 at 16.04.09@2x.png…]()

**Reply by mhenrixon** [reply]
```ruby
cosmos(dev)> ApplicationComponent.new.method(:capture)
=> #<Method: ApplicationComponent(ActionView::Helpers::CaptureHelper)#capture(*args, &block) /Users/mhenrixon/.gem/ruby/3.3.0/bundler/gems/rails-979c6b0366c0/actionview/lib/action_view/helpers/capture_helper.rb:47>
cosmos(dev)>
```

Looks like actionview has a capture helper method @joeldrapper

**Reply by mhenrixon** [reply]
So unless you rename your capture method, we can't mix ERB and Phlex, which, in my case, makes the upgrade hard as I can't roll it out gradually.

**Reply by joeldrapper** [reply]
Ah, this explains it! Why do you need Rails’ capture method in a Phlex component? How did it get in?

**Reply by mhenrixon** [reply]
That's what I just wrote you. I am using an ERB layout, so Phlex and ERB currently do not mix, which makes it hard to roll this out gradually, @joeldrapper!

**Reply by mhenrixon** [reply]
```ruby
cosmos(dev)> Module.constants.select { |c| (eval c.to_s).is_a?(Class) && (eval c.to_s).include?(ActionView::Helpers::CaptureHe
lper)}
=>
[:ApplicationView,
 :ApplicationLayout,
 :ConsultingLayout,
 :ApplicationComponent,
 :AvatarComponent,
 :FlashComponent,
 :FooterComponent,
 :MenuComponent,
 :ModalComponent,
 :NotificationComponent,
 :NotificationTwoComponent,
 :SubmenuComponent,
 :TagComponent,
 :TimelineComponent,
 :NotificationTwoComponentPreview,
 :NotificationComponentPreview]
```

Hmm how is this even possible?

**Reply by mhenrixon** [reply]
Aha! I found it! https://github.com/abeidahmed/rails-heroicon/blob/main/lib/rails_heroicon/helper.rb I will write a specific one for my scenario with phlex.

**Reply by joeldrapper** [reply]
> I am using an ERB layout, so Phlex and ERB currently do not mix

Using an ERB layout should be fine. It shouldn't mean Phlex::HTML#capture gets overridden on Phlex::HTML objects.

> Aha! I found it!

I don't see the capture definition in there. 🤔 Is it included in `ActionView::Helpers::TagHelper`?

I don't know if it's still maintained but you may find this project useful https://github.com/nejdetkadir/phlex-heroicons

**Reply by mhenrixon** [reply]
Awesome! I fixed it for now; I added a generic heroicon component. I'm still trying to figure out how I feel about that gem. It needs to be more terse for something like a heroicon. 

I added a method called `_heroicon` to go with my existing `_` method for finding components. 

Still, I appreciate the suggestion, and I'll look into the code in more detail tomorrow. I also encountered a problem where I originally had the method as `heroicon,` which was not working in ERB. 

The joys of migrations. I thank thee for thyne assistance, I understand phlex much better now.
